docs/middleware.md

```diff
@@ -232,6 +232,252 @@ around explicitly, rather than mutating the middleware instance.
+
+### Wrapping the request
+
+On Starlette, we have data structures that make it more convenient to organize your ASGI middleware.
+
+For example, we can create a `Request` object, and work with it.
+```python
+from starlette.requests import Request
+
+class ASGIMiddleware:
+    def __init__(self, app):
+        self.app = app
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] == "http":
+            request = Request(scope, receive, send)
+            # Do something here!
+        await self.app(scope, receive, send)
+```
+
+Also, it's possible to make use of `MutableHeaders` to change the headers:
+
+```python
+class ASGIMiddleware:
+    def __init__(self, app, headers):
+        self.app = app
+        self.headers = headers
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] == "http":
+            async def send_wrapper(message):
+                if message["type"] == "http.response.start":
+                    headers = MutableHeaders(scope=message)
+                    for key, value for self.headers:
+                        headers.append(key, value)
+                await send(message)
+        await self.app(scope, receive, send_wrapper)
+```
+
+### Per-request state
+
+The ASGI middleware class is meant to be stateless, as we don't want to spread the state from one request to another.
+The responder pattern is a way to deal with per-request state. The idea is to have an object that will hold that information.
+
+```python
+class ASGIMiddleware:
+    def __init__(self, app):
+        self.app = app
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] == "http":
+            responder = Responder(self.app)
+            return await responder(scope, receive, send)
+
+        await self.app(scope, receive, send)
+
+
+class Responder:
+    def __init__(self, app):
+        self.app = app
+        self.started = False
+
+    def __call__(self, scope, receive, send):
+        async def send_wrapper(message):
+            if message["type"] == "http.response.body" and not self.started:
+                self.started = True
+            await send(message)
+        await self.app(scope, receive, send_wrapper)
+```
+
+The `Responder` object, in this case, will save the information about when the response body has "started" to be sent.
+
+!!! info
+    You can see a more advanced implementation of this pattern on our [GZipMiddleware](), and also on the [msgpack-asgi](https://github.com/florimondmanca/msgpack-asgi/blob/6261b1e22b3689551f68038ee00adda5fbc04670/src/msgpack_asgi/_middleware.py).
+
+### Storing context in `scope`
+
+As we know by now, the `scope` holds the information about the application. To be precise, the `scope` holds
+the stateless information of the application. IS THIS CORRECT?
+
+As per the ASGI specifications, any application can store custom information on the `scope`.
+To be precise, it should be stored under the `extensions` key.
```

Comment:
> As per the ASGI specifications, any application can store custom information on the `scope`.
> To be precise, it should be stored under the `extensions` key.

Not really sure about this. As per https://github.com/encode/starlette/discussions/1511 it seems like storing non-ASGI state in `scope` is not really spec'd yet.

That's also different from [ASGI extensions](https://asgi.readthedocs.io/en/latest/extensions.html) which are meant to convey information from and to the ASGI server (eg. Uvicorn).

So the current status seems to be:

* Middleware may conveniently add items to `scope` to make data available to endpoints.
* Make sure to properly namespace to avoid conflicts.
* This is mostly a convention -- not standardized yet.

So, _maybe_ the example could be more like this:

```python
scope["my-special-feature"] = {"enabled": True}
```

But it's not even clear if we _should_ do that _or_ edit `extensions` (which we'd need to `.setdefault()` since it might already be set due to other middleware...?). I think it's a safer, "sure not to interfere with other middleware" option.

Reply:
> Not really sure about this. As per https://github.com/encode/starlette/discussions/1511 it seems like storing non-ASGI state in scope is not really spec'd yet.

Right. 🤔 

Thanks for this comment. I want to have that clarified by the ASGI specs as well. I've created an issue there, and cc you.

Reply:
I'm going to change this according to the asgiref issue mentioned above.

Reply:
I agree on *not* using the `extensions` word here. I would expect extensions to be used for generic/generalizable things that could become a standard in some way, and as Florimond says, that interact with the server (Uvicorn) directly in some way.

Also, I would expect this guide to be a bit more directed to users creating a middleware for their own app, not a generic/generalizable thing that could become a standard. I would just remove the `extensions` word and say something around trying to avoid collisions in names.

```suggestion
As per the ASGI specifications, any application can store custom information on the `scope`.
Have in mind that other components could also store data in the same `scope`, so it's important to use a key that has a low chance of being used by other things.

For example, if you are building an application called `supper-app` you could have that as a prefix for any keys you put in the `scope`, and then you could have a key called `super-app-transaction-id`.
```